Convert the EQ IIR component to use the module interface

ranj063 opened the pull request.

marc-hb commented:
Is it possible to fix the regression #5911 introduced by #5766 first?

ranj063 commented:
> Is it possible to fix the regression #5911 introduced by #5766 first?

this PR is not related to that issue at all. But I have a potential fix for it

singalsu commented on src/audio/eq_iir/eq_iir.c:
```diff
@@ -125,6 +131,9 @@ static void eq_iir_s24_default(const struct comp_dev *dev, const struct audio_st
 	int processed = 0;
 
+	bsource->consumed += samples << 2;
+	bsink->size += samples << 2;
```
Same

singalsu commented on src/audio/eq_iir/eq_iir.c:
```diff
@@ -171,6 +182,9 @@ static void eq_iir_s32_default(const struct comp_dev *dev, const struct audio_st
 	int processed = 0;
 
+	bsource->consumed += samples << 2;
+	bsink->size += samples << 2;
```
and here

singalsu commented on src/audio/eq_iir/eq_iir.c:
```diff
@@ -216,6 +233,9 @@ static void eq_iir_s32_16_default(const struct comp_dev *dev, const struct audio
 	int processed = 0;
 
+	bsource->consumed += samples << 2;
+	bsink->size += samples << 1;
```
This is more special so, separate macros of functions for source and sink when different format.

singalsu commented on src/audio/eq_iir/eq_iir.c:
```diff
@@ -286,27 +312,39 @@ static void eq_iir_s32_24_default(const struct comp_dev *dev, const struct audio
+	} else {
+		bsource->consumed += (frames * source->channels) << 2;
+		bsink->size += (frames * source->channels) << 2;
+	}
```
Maybe all of this could happen in one function since many components have  a pass-thru mode.

singalsu commented on src/audio/eq_iir/eq_iir.c:
```diff
@@ -699,236 +734,165 @@ static int eq_iir_verify_params(struct comp_dev *dev,
+ */
+static void eqiir_set_alignment(struct audio_stream *source, struct audio_stream *sink)
+{
+#if XCHAL_HAVE_HIFI3 || XCHAL_HAVE_HIFI4
```
This HiFi3 version difference is not yet needed but I think it can be left this way since @andrula-song will implement a more efficient version of IIR in Q3.

singalsu reviewed (commented):
Looks quite good to me. Consider some improvements those I proposed to clean up this and successive components.

fredoh9 commented:
I quickly checked this PR on my ADL RVP which has a DMIC transducer board. Captured wavefile is fine also

ranj063 commented:
> Looks quite good to me. Consider some improvements those I proposed to clean up this and successive components.

@singalsu I've addressed your comments in this version now

ranj063 commented on src/include/sof/audio/audio_stream.h:
```diff
@@ -800,6 +800,39 @@ static inline void audio_stream_fmt_conversion(enum ipc4_bit_depth depth,
+	const uint32_t frame_align_req = 1;
+
+#endif
+
```
@andrula-song it looks like the volume and both the eqiir and fir can use this function for now. If other components have different requirements, we'd just need to use a different alignment than this default one no?

andrula-song replied:
Then how about renaming this function like audio_stream_set_alignment_default to avoid misunderstanding?

singalsu replied:
Can this be moved back to be volume specific and use as default the generic align?

lyakh commented:
@ranj063 this PR now has conflicts after my cache / sparse PR with a patch for eq-iir has been merged, please update

singalsu reviewed (approved):
Looks good now, but need the rebase.

lgirdwood commented:
@ranj063 fwiw, All the sparse cache component updates now merged.

lgirdwood commented:
@ranj063 ping

ranj063 commented:
> @ranj063 ping

@lgirdwood updated now

lgirdwood commented on test/cmocka/src/audio/eq_iir/eq_iir_process.c:
```diff
@@ -69,39 +71,51 @@ static struct sof_ipc_comp_process *create_eq_iir_comp_ipc(struct test_data *td)
 	size_t ipc_size = sizeof(struct sof_ipc_comp_process);
 	struct sof_abi_hdr *blob = (struct sof_abi_hdr *)iir_coef_2ch;
+	uint8_t uuid[SOF_UUID_SIZE] = {
+		230, 192, 80, 81, 249, 39, 200, 78, 131, 81, 199, 5, 182, 66, 209, 47};
```
I think all the UUID values are in hex in other places

lgirdwood commented on src/include/sof/audio/module_adapter/module/generic.h:
```diff
@@ -334,4 +334,28 @@ int module_set_large_config(struct comp_dev *dev, uint32_t param_id, bool first_
 			    bool last_block, uint32_t *data_offset, char *data);
 
+static inline void
+update_module_buffers(struct input_stream_buffer *input, struct output_stream_buffer *output,
```
```module_update_processing_position()``` ?

lgirdwood reviewed (commented):
Just a few minor questions.

keqiaozhang commented:
SOFCI TEST

lgirdwood commented:
SOFCI TEST

lgirdwood commented:
rerun CI, there has been a west issue.

lyakh commented on src/audio/eq_iir/eq_iir.c:
```diff
@@ -704,250 +734,158 @@ static int eq_iir_verify_params(struct comp_dev *dev,
-		ret = eq_iir_setup(cd, source_c->stream.channels);
+		ret = eq_iir_setup(mod, source_c->stream.channels);
+		buffer_release(sink_c);
+		buffer_release(source_c);
```
you now have 3 locations where you release buffers. Wouldn't it be better to keep the original `if {} else {}` code structure and release the buffers once at the end?

marc-hb commented:
This PR did not pass PR CI: https://sof-ci.01.org/sofpr/PR5926/build903/devicetest/

Unsurprisingly, there are now a lot of failures in daily test  run 14058
Start Time: 2022-07-18 21:29:41 UTC
End Time: 2022-07-19 02:32:40 UTC
Kernel Branch: topic/sof-dev
Kernel Commit: [c02c3b653c73 (https://github.com/thesofproject/linux/commit/c02c3b653c73)
SOF Branch: main
SOF Commit: [ea72e3e97e10](https://github.com/thesofproject/sof/commit/ea72e3e97e10)

[SOF#6020:](https://github.com/thesofproject/sof/issues/6020) [BUG] DSP panic when testing signal stop/start on DMIC pipeline 
[SOF#6022](https://github.com/thesofproject/sof/issues/6022): [BUG][APL_UP2_PCM512X] kcontrol xx read failed for widget EQIIR y.z

(Thanks @keqiaozhang for the identification)

ranj063 commented:
> This PR did not pass PR CI: https://sof-ci.01.org/sofpr/PR5926/build903/devicetest/
> 
> Unsurprisingly, there are now a lot of failures in daily test run 14058 Start Time: 2022-07-18 21:29:41 UTC End Time: 2022-07-19 02:32:40 UTC Kernel Branch: topic/sof-dev Kernel Commit: [c02c3b653c73 (https://github.com/thesofproject/linux/commit/c02c3b653c73) SOF Branch: main SOF Commit: [ea72e3e97e10](https://github.com/thesofproject/sof/commit/ea72e3e97e10)
> 
> [SOF#6020:](https://github.com/thesofproject/sof/issues/6020) [BUG] DSP panic when testing signal stop/start on DMIC pipeline [SOF#6022](https://github.com/thesofproject/sof/issues/6022): [BUG][APL_UP2_PCM512X] kcontrol xx read failed for widget EQIIR y.z
> 
> (Thanks @keqiaozhang for the identification)

Thanks @marc-hb looking at this now

marc-hb commented:
Temporary revert submitted in #6029, thx @ranj063
